cmd/charm/charmcmd/plugin.go

```diff
@@ -97,116 +98,235 @@ func pluginHelpTopic() string {
+		resultDescriptionMap[result.Name] = result
 		helpResult := <-help
-		resultHelpMap[helpResult.name] = helpResult
+		resultHelpMap[helpResult.Name] = helpResult
 	}
 	// plugins array is already sorted, use this to get the results in order.
 	for _, plugin := range plugins {
 		// Strip the 'charm-' off the start of the plugin name in the results.
-		result := resultDescriptionMap[plugin]
-		result.name = result.name[len(pluginPrefix):]
-		result.doc = resultHelpMap[plugin].doc
+		result := resultDescriptionMap[plugin.name]
+		result.Name = result.Name[len(pluginPrefix):]
+		result.Doc = resultHelpMap[plugin.name].Doc
 		results = append(results, result)
 	}
 	pluginDescriptionsResults = results
+
+	writeResultsCache(pluginCache, results)
 	return results
 }
 
+func readAndReturnCacheIfValid(pluginCache string, plugins []fileInfo, pluginExists map[string]int) []pluginDescription {
+	results := []pluginDescription{}
+	if f, err := os.Open(pluginCache); err == nil {
+		decoder := json.NewDecoder(f)
+		if err = decoder.Decode(&results); err == nil {
+			cachedExists := map[string]int{}
+		compare:
+			for _, cachedPlugin := range results {
+				cachedExists[pluginPrefix+cachedPlugin.Name]++
+				// If a plugin no longer exists in cache, invalidate entire cache.
+				if pluginExists[pluginPrefix+cachedPlugin.Name] == 0 {
+					results = nil
+					break compare
+				}
+				// Compare ModTime of each found plugin to its cached plugin.
+				for _, newp := range plugins {
+					if newp.name == pluginPrefix+cachedPlugin.Name {
+						filename := filepath.Join(newp.dir, newp.name)
+						stat, err2 := os.Stat(filename)
+						if err2 != nil {
+							logger.Errorf("could not stat %s", filename, err2)
+							results = nil
+							break compare
+						}
+						mtime, err2 := strconv.ParseInt(cachedPlugin.ModTime, 0, 64)
+						if stat.ModTime().Unix() != mtime {
+							// We could invalidate just this plugin, but it is a rare occurance so invalidate all.
+							results = nil
+							break compare
+						}
+					}
+				}
+			}
+			for _, newp := range plugins {
+				// If the cache is missing a found plugin, invalidate entire cache.
+				if cachedExists[newp.name] == 0 {
+					results = nil
+				}
+			}
+
+			// The cache was not invalidated for any reason, so use it.
+			if results != nil {
+				return results
+			}
+		} else {
+			logger.Errorf("There was a problem decoding json %s", err)
+		}
+	}
+	return nil
+}
+
+func writeResultsCache(pluginCache string, results []pluginDescription) {
+	if f, err := os.Create(pluginCache); err == nil {
+		encoder := json.NewEncoder(f)
+		if err = encoder.Encode(results); err != nil {
+			logger.Errorf("encoding cached plugin descriptions: %s", err)
+		}
+	} else {
+		logger.Errorf("opening plugin cache file: %s", err)
+	}
+}
+
 type pluginDescription struct {
-	name        string
-	description string
-	doc         string
+	Name        string
+	Description string
+	Doc         string
+	ModTime     string
```

Comment:
Good catch, I had that, but with lstat v. stat I thought it was a JS round trip issue so moved to string. I was wrong and have moved this back to time.Time.